Objective-specific equipment will always be given.

Y0SH1M4S73R opened the pull request:
## About The Pull Request

#60686 was caused by `/objective/give_special_equipment()` qdelling the special equipment if it was unable to put it in the antag's backpack. This PR allows antagonists to use their uplink to get the equipment they couldn't receive (or call in a pod, if they don't have an uplink)

## Why It's Good For The Game

Fixes #60686

## Changelog
:cl:
fix: Traitors that are unable to receive equipment required for specific objectives will be able to order that equipment from the uplink. Other antagonists that are unable to receive such equipment can call in a pod containing that equipment, unless they somehow have an unlocked uplink on their person, in which case they can order it from the uplink.
/:cl:

Y0SH1M4S73R committed:
special equipment will always be given

oklews commented:
It would be very weird to see it randomly drop down in the middle of the floor of some social situation. Maybe instead give them a warning or something before it appears?

Y0SH1M4S73R commented:
I was thinking about warning the user that they don't have room for some of the equipment, then calling `give_special_equipment` again after a few seconds; however, this could be spammy, especially if multiple objectives have special equipment.

GuillaumePrata commented:
I thought this issue was fixed already.
I know it's out of scope for this PR but If spawning with multiple traitor tools on the floor become an issue, maybe it would be the time to consider putting them as a 0 TC purchase because at least I always hate when I arrest someone for something unrelated and find unused SM stealing tools in their bag, it's such a lame interaction.

Mothblocks reviewed (changes requested):
Antag stuff dropping on the ground is definitely a no-no. I was gonna suggest an action button to drop it when you want, but having it as a 0-TC item is a better idea.

jughu commented:
I wouldn't make it a 0 tc item 5 tc item because failure must be punishing if you lose the orginal kit

Cyberboss commented:
> 
> 
> I wouldn't make it a 0 tc item 3 tc item because failure must be punishing if you lose the orginal kit

0TC, 1 stock

Y0SH1M4S73R commented:
> 
> 
> I wouldn't make it a 0 tc item 3 tc item because failure must be punishing if you lose the orginal kit

This isn't supposed to be for if you lose the original kit - this is for when the kit can't be put in your inventory in the first place.

YakumoChen commented:
can't changelings also get danger-thing steal objectives?

Y0SH1M4S73R commented:
> 
> 
> can't changelings also get danger-thing steal objectives?
> 
> Edit: yeah, they can. It needs to be accessible to any antag without an uplink.

Maybe the fallback could be a single-use implant that spawns the equipment.

YakumoChen commented:
Just make spawning it a one time action button that attempts to put it in your bag/hands/the ground

Mothblocks commented:
I proposed the action button but decided against it since I think a 0TC item is a lot better

Mothblocks commented:
Do they even?

I'd use an action button as a fallback in those cases, but there's some reasons why if an uplink can be used it should, mostly just that it's weird that you can be totally stripped and still spawn some SM extracting tools from your ass.

YakumoChen commented:
implant?
Make it pod down on request?

jughu commented:
What if you make the box 3 tc and if you have the objective you get a 100% discount on it

YakumoChen commented:
> 
> 
> What if you make the box 3 tc and if you have the objective you get a 100% discount on it

not sure we should enable any traitor to disable, say, a Nukeops nuke on demand without the objective.

Y0SH1M4S73R committed:
reworks fallback to use uplinks when possible

and a self-deleting spell, when not possible

Y0SH1M4S73R committed:
moves some proc order around

this prevents traitors from getting the pod-calling spell when they can buy from their uplink anyways

Y0SH1M4S73R committed:
accounts for the case of a traitor specifically not being given an uplink but having an unlocked one on them

Y0SH1M4S73R committed:
Merge branch 'master' into midround_sliver_objective_fix

Mothblocks commented on code/datums/components/uplink.dm:
```diff
@@ -93,6 +94,23 @@
 
+/datum/component/uplink/proc/update_special_equipment(mob/user, updated_items)
+	if(user?.mind?.failed_special_equipment)
+		for(var/eq_path in user.mind.failed_special_equipment)
+			var/obj/item/eq_prototype = eq_path
```
```suggestion
		for(var/obj/item/equipment_path as anything in user.mind.failed_special_equipment)
```

And don't shorten `equipment` to `eq`.

Mothblocks commented on code/datums/mind.dm:
```diff
@@ -696,13 +698,27 @@
+					continue
+			return ret
+
+/datum/mind/proc/try_give_equipment_fallback()
```
Doc comment.

Mothblocks commented on code/datums/mind.dm:
```diff
@@ -696,13 +698,27 @@
-		. = I.GetComponent(/datum/component/uplink)
-		if(.)
-			break
+		var/datum/component/uplink/ret = I.GetComponent(/datum/component/uplink)
```
Better variable name.

Mothblocks replied:
And not "return_value" or anything like that, simply just `found_uplink` is fine.

Mothblocks commented on code/datums/mind.dm:
```diff
@@ -696,13 +698,27 @@
-			break
+		var/datum/component/uplink/ret = I.GetComponent(/datum/component/uplink)
+		if(ret)
+			if(check_unlocked)
+				if(ret.locked)
+					continue
```
Combine into one.

Mothblocks commented on code/datums/mind.dm:
```diff
@@ -696,13 +698,27 @@
+	if(traitor_datum)
+		uplink = traitor_datum.uplink
+	if(!uplink)
+		uplink = find_syndicate_uplink(TRUE)
```
```suggestion
		uplink = find_syndicate_uplink(check_unlocked = TRUE)
```

Mothblocks commented on code/datums/mind.dm:
```diff
@@ -696,13 +698,27 @@
+		uplink = traitor_datum.uplink
+	if(!uplink)
+		uplink = find_syndicate_uplink(TRUE)
+	if(!uplink)
+		if(!(locate(/obj/effect/proc_holder/spell/self/special_equipment_fallback) in spell_list))
```
Combine these two with `&&`

Mothblocks commented on code/game/gamemodes/objective.dm:
```diff
@@ -159,11 +159,39 @@ GLOBAL_LIST_EMPTY(objectives)
 			for(var/eq_path in special_equipment)
-				var/obj/O = new eq_path
-				H.equip_in_one_of_slots(O, slots)
+				var/obj/object = new eq_path
```
1. Use `as anything` in the loop rather than reassigning it here.
2. This is a useless variable name, what object is it?

Y0SH1M4S73R replied:
`object` is a concrete instance of `eq_path`. If I were to simply say `eq_path = new eq_path`, then `equip_in_one_of_slots` will qdel `eq_path`, resulting in a `null` getting added to `failed_special_equipment`. For clarity's sake, I will however rename `eq_path` to `equipment_path` and `object` to `equipment_object`

Mothblocks replied:
I didn't notice the `new`, you're right.

Mothblocks commented on code/modules/uplink/uplink_items.dm:
```diff
@@ -2035,3 +2035,16 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	illegal_tech = FALSE
+
+// Special equipment (Dynamically fills in uplink component)
+/datum/uplink_item/special_equipment
```
Have we confirmed that, unless you're in the specific case where you can't get an item, this won't show?

Y0SH1M4S73R replied:
Yes, I have tested it on a local branch.

Y0SH1M4S73R committed:
addresses review items

Y0SH1M4S73R committed:
Update objective.dm

Mothblocks commented on code/datums/components/uplink.dm:
```diff
@@ -93,6 +94,22 @@
 			updated_items[category] = uplink_items[category]
 
+/datum/component/uplink/proc/update_special_equipment(mob/user, updated_items)
+	if(user?.mind?.failed_special_equipment)
```
Early-return.

Mothblocks reviewed (changes requested):
Last nitpick.

Y0SH1M4S73R committed:
addresses one final nitpick
